File: components/PageControl/examples/PageControlTypicalUseExample.m
//
//  PageControlTypicalUseExample.m
//  GTCatalog
//
//  Created by liuxc on 2018/8/28.
//
#import <UIKit/UIKit.h>

#import "GTPageControl.h"

@interface PageControlTypicalUseExample : UIViewController <UIScrollViewDelegate>
@end

@implementation PageControlTypicalUseExample{
    UIScrollView *_scrollView;
    GTCPageControl *_pageControl;
    NSArray *_pages;
}


- (void)viewDidLoad {
    [super viewDidLoad];
    CGRect standardizedFrame = CGRectStandardize(self.view.frame);
    CGFloat boundsWidth = CGRectGetWidth(standardizedFrame);
    CGFloat boundsHeight = CGRectGetHeight(standardizedFrame);

    NSArray *pageColors = @[
        [UIColor colorWithWhite:0.9f alpha:1.0f],
        [UIColor colorWithWhite:0.8f alpha:1.0f],
        [UIColor colorWithWhite:0.7f alpha:1.0f],
    ];

    // Scroll view configuration
    _scrollView = [[UIScrollView alloc] initWithFrame:self.view.bounds];
    _scrollView.autoresizingMask = UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleHeight;
    _scrollView.delegate = self;
    _scrollView.pagingEnabled = YES;
    _scrollView.contentSize = CGSizeMake(boundsWidth * pageColors.count, boundsHeight);
    _scrollView.showsHorizontalScrollIndicator = NO;
    [self.view addSubview:_scrollView];

    NSMutableArray *pages = [NSMutableArray array];

    // Add pages to scrollView.
    for (NSUInteger i = 0; i < pageColors.count; i++) {
        CGRect pageFrame = CGRectOffset(self.view.bounds, i * boundsWidth, 0);
        UILabel *page = [[UILabel alloc] initWithFrame:pageFrame];
        page.text = [NSString stringWithFormat:@"Page %lu", (unsigned long)i + 1];
        page.font = [UIFont systemFontOfSize:50];
        page.textColor = [UIColor colorWithWhite:0 alpha:0.8f];
        page.textAlignment = NSTextAlignmentCenter;
        page.backgroundColor = pageColors[i];
        page.autoresizingMask =
        UIViewAutoresizingFlexibleTopMargin | UIViewAutoresizingFlexibleBottomMargin;
        [_scrollView addSubview:page];
        [pages addObject:page];
    }
    _pages = [pages copy];

    // Page control configuration.
    _pageControl = [[GTCPageControl alloc] initWithFrame:CGRectZero];
    _pageControl.numberOfPages = pageColors.count;

    [_pageControl addTarget:self
                     action:@selector(didChangePage:)
           forControlEvents:UIControlEventValueChanged];
    _pageControl.autoresizingMask =
    UIViewAutoresizingFlexibleTopMargin | UIViewAutoresizingFlexibleWidth;
    [self.view addSubview:_pageControl];

    [self.view setNeedsLayout];

}

#pragma mark - Frame changes

- (void)viewWillLayoutSubviews {
    [super viewWillLayoutSubviews];
    NSInteger pageBeforeFrameChange = _pageControl.currentPage;
    NSInteger pageCount = _pages.count;
    CGRect standardizedFrame = CGRectStandardize(self.view.frame);
    for (NSInteger i = 0; i < pageCount; i++) {
        UILabel *page = [_pages objectAtIndex:i];
        page.frame =
        CGRectOffset(self.view.bounds, i * CGRectGetWidth(standardizedFrame), 0);
    }
    _scrollView.contentSize =
    CGSizeMake(CGRectGetWidth(standardizedFrame) * pageCount, CGRectGetHeight(standardizedFrame));
    CGPoint offset = _scrollView.contentOffset;
    offset.x = pageBeforeFrameChange * CGRectGetWidth(standardizedFrame);
    // This non-anmiated change of offset ensures we keep the same page
    [_scrollView setContentOffset:offset animated:NO];
    _scrollView.frame = self.view.bounds;

    // We want the page control to hug the bottom of the screen.
    UIEdgeInsets edgeInsets = UIEdgeInsetsZero;
#if defined(__IPHONE_11_0) && (__IPHONE_OS_VERSION_MAX_ALLOWED >= __IPHONE_11_0)
    if (@available(iOS 11.0, *)) {
        // Accommodate insets for iPhone X.
        edgeInsets = self.view.safeAreaInsets;
    }
#endif
    [_pageControl sizeToFit];
    CGFloat yOffset =
    CGRectGetHeight(self.view.frame) - CGRectGetHeight(_pageControl.frame) - edgeInsets.bottom;
    _pageControl.frame =
    CGRectMake(0, yOffset, CGRectGetWidth(self.view.frame), CGRectGetHeight(_pageControl.frame));

}

#pragma mark - UIScrollViewDelegate

- (void)scrollViewDidScroll:(UIScrollView *)scrollView {
    [_pageControl scrollViewDidScroll:scrollView];
}

- (void)scrollViewDidEndDecelerating:(UIScrollView *)scrollView {
    [_pageControl scrollViewDidEndDecelerating:scrollView];
}

- (void)scrollViewDidEndScrollingAnimation:(UIScrollView *)scrollView {
    [_pageControl scrollViewDidEndScrollingAnimation:scrollView];
}

#pragma mark - User events

- (void)didChangePage:(GTCPageControl *)sender {
    CGPoint offset = _scrollView.contentOffset;
    offset.x = sender.currentPage * _scrollView.bounds.size.width;
    [_scrollView setContentOffset:offset animated:YES];
}

#pragma mark - GTCatalog

+ (NSArray *)catalogBreadcrumbs {
    return @[ @"Page Control", @"Page Control" ];
}

+ (NSString *)catalogDescription {
    return @"This control is designed to be a drop-in replacement for UIPageControl, with a user"
    " experience influenced by Material Design.";
}

+ (BOOL)catalogIsPrimaryDemo {
    return YES;
}

+ (BOOL)catalogIsPresentable {
    return YES;
}

@end

File: components/PageControl/examples/PageControlTypicalUseExample.swift
//
//  PageControlTypicalUseExample.swift
//  GTCatalog
//
//  Created by liuxc on 2018/8/28.
//

import UIKit

import GTKitComponents.GTPageControl
import GTKitComponents.GTPalettes

class PageControlSwiftExampleViewController: UIViewController, UIScrollViewDelegate {

    static let pageColors = [
        GTCPalette.cyan.tint300,
        GTCPalette.cyan.tint500,
        GTCPalette.cyan.tint700,
        GTCPalette.cyan.tint300,
        GTCPalette.cyan.tint500,
        GTCPalette.cyan.tint700
    ]

    let pageControl = GTCPageControl()
    let scrollView = UIScrollView()
    let pageLabels: [UILabel] = PageControlSwiftExampleViewController.pageColors.enumerated().map {
        enumeration in
        let (i, pageColor) = enumeration
        let pageLabel = UILabel()
        pageLabel.text = "Page \(i + 1)"
        pageLabel.font = pageLabel.font.withSize(50)
        pageLabel.textColor = UIColor(white: 0, alpha: 0.8)
        pageLabel.backgroundColor = pageColor
        pageLabel.textAlignment = NSTextAlignment.center
        pageLabel.autoresizingMask = [.flexibleTopMargin, .flexibleBottomMargin]
        return pageLabel
    }

    override func viewDidLoad() {
        super.viewDidLoad()
        view.backgroundColor = UIColor.white

        scrollView.frame = self.view.bounds
        scrollView.autoresizingMask = [.flexibleWidth, .flexibleHeight]
        scrollView.delegate = self
        scrollView.isPagingEnabled = true
        scrollView.contentSize = CGSize(width: view.bounds.width * CGFloat(pageLabels.count),
                                        height: view.bounds.height)
        scrollView.showsHorizontalScrollIndicator = false
        view.addSubview(scrollView)

        // Add pages to scrollView.
        for (i, pageLabel) in pageLabels.enumerated() {
            let pageFrame = view.bounds.offsetBy(dx: CGFloat(i) * view.bounds.width, dy: 0)
            pageLabel.frame = pageFrame
            scrollView.addSubview(pageLabel)
        }

        pageControl.numberOfPages = pageLabels.count

        pageControl.addTarget(self, action: #selector(didChangePage), for: .valueChanged)
        pageControl.autoresizingMask = [.flexibleTopMargin, .flexibleWidth]
        view.addSubview(pageControl)
    }

    // MARK: - Frame changes

    override func viewWillLayoutSubviews() {
        super.viewWillLayoutSubviews()
        let pageBeforeFrameChange = pageControl.currentPage
        for (i, pageLabel) in pageLabels.enumerated() {
            pageLabel.frame = view.bounds.offsetBy(dx: CGFloat(i) * view.bounds.width, dy: 0)
        }
        scrollView.contentSize = CGSize(width: view.bounds.width * CGFloat(pageLabels.count),
                                        height: view.bounds.height)
        var offset = scrollView.contentOffset
        offset.x = CGFloat(pageBeforeFrameChange) * view.bounds.width
        // This non-anmiated change of offset ensures we keep the same page
        scrollView.contentOffset = offset

        var edgeInsets = UIEdgeInsets.zero;
        #if swift(>=3.2)
        if #available(iOS 11, *) {
            edgeInsets = self.view.safeAreaInsets
        }
        #endif
        let pageControlSize = pageControl.sizeThatFits(view.bounds.size)
        let yOffset = self.view.bounds.height - pageControlSize.height - 8 - edgeInsets.bottom;
        pageControl.frame =
            CGRect(x: 0, y: yOffset, width: view.bounds.width, height: pageControlSize.height)
    }

    // MARK: - UIScrollViewDelegate

    func scrollViewDidScroll(_ scrollView: UIScrollView) {
        pageControl.scrollViewDidScroll(scrollView)
    }

    func scrollViewDidEndDecelerating(_ scrollView: UIScrollView) {
        pageControl.scrollViewDidEndDecelerating(scrollView)
    }

    func scrollViewDidEndScrollingAnimation(_ scrollView: UIScrollView) {
        pageControl.scrollViewDidEndScrollingAnimation(scrollView)
    }

    // MARK: - User events

    @objc func didChangePage(_ sender: GTCPageControl) {
        var offset = scrollView.contentOffset
        offset.x = CGFloat(sender.currentPage) * scrollView.bounds.size.width
        scrollView.setContentOffset(offset, animated: true)
    }

    // MARK: - CatalogByConvention

    @objc class func catalogBreadcrumbs() -> [String] {
        return [ "Page Control", "Swift example"]
    }

    @objc class func catalogIsPrimaryDemo() -> Bool {
        return false
    }
}
